### ISSUE TITLE
imfile not reading/skipping last line?

### ISSUE THREAD

**Issue by mostolog** [rsyslog/rsyslog#1248]
Hi

As detected while testing https://github.com/rsyslog/rsyslog/issues/498, imfile seems not reading/sending last line.
Attached example with 5 lines [example.log.txt](https://github.com/rsyslog/rsyslog/files/575052/example.log.txt) but sending 4.
I'm using 8.22

```
...
4298.009984338:action 0 queue:Reg/w0: frame to send: '2 syslog 637 redacted {"file":"/logs/apps/app/app.log","msg":"INFO class:000 - User - Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum."}
'
4298.010000522:action 0 queue:Reg/w0: relpTcpSend: sock 8, lenbuf 651, send returned 651 [errno 11]
4298.010006268:action 0 queue:Reg/w0: ADD sess 0x7f6890003e90 unacked 1, sessState 4
4298.010010845:action 0 queue:Reg/w0: sendbuf added to unacked list
4298.010017879:action 0 queue:Reg/w0: omrelp: endTransaction
4298.010059234:action 0 queue:Reg/w0: Action 0 transitioned to state: rdy
4298.010066603:action 0 queue:Reg/w0: actionCommit, action 0, in retry loop, iRet 0
4298.010072349:action 0 queue:Reg/w0: regular consumer finished, iret=0, szlog 3 sz phys 4
4298.010092162:action 0 queue:Reg/w0: queue.c:1947:ConsumerReg:invocation 3: WARNING: mutex still owned by us as we exit function, mutex:  0x1a87fa0[0/action 0 queue:Reg/w0 (7f689ac8e700)]
4298.010116054:action 0 queue:Reg/w0: DeleteProcessedBatch: we deleted 1 objects and enqueued 0 objects
4298.010124335:action 0 queue:Reg/w0: doDeleteBatch: delete batch from store, new sizes: log 3, phys 3
4298.010159950:action 0 queue:Reg/w0: action 'action 0': is transactional - executing in commit phase
4298.010166104:action 0 queue:Reg/w0: omrelp: beginTransaction
4298.010171559:action 0 queue:Reg/w0: Action 0 transitioned to state: itx
4298.010189981:action 0 queue:Reg/w0: action 'action 0': is transactional - executing in commit phase
4298.010215841:action 0 queue:Reg/w0: action 'action 0': is transactional - executing in commit phase
4298.010222658:action 0 queue:Reg/w0: doTransaction: action 0, currIParam 3
4298.010228109:action 0 queue:Reg/w0: omrelp: shutdownImmediate ptr now is 0x1a81bd0
4298.010233032:action 0 queue:Reg/w0: entering actionCalldoAction(), state: itx, actionNbr 0
4298.010238091:action 0 queue:Reg/w0:  redacted:20514/RELP
4298.010243208:action 0 queue:Reg/w0: relpSessWaitState returns 0
4298.010251159:action 0 queue:Reg/w0: frame to send: '3 syslog 1084 redacted {"file":"/logs/apps/app/app.log","msg":"INFO class:000 - User - Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum. Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum."}
'
4298.010266810:action 0 queue:Reg/w0: relpTcpSend: sock 8, lenbuf 1099, send returned 1099 [errno 11]
4298.010271977:action 0 queue:Reg/w0: ADD sess 0x7f6890003e90 unacked 2, sessState 4
4298.010276351:action 0 queue:Reg/w0: sendbuf added to unacked list
4298.010282107:action 0 queue:Reg/w0: omrelp: shutdownImmediate ptr now is 0x1a81bd0
4298.010286815:action 0 queue:Reg/w0: entering actionCalldoAction(), state: itx, actionNbr 0
4298.010291576:action 0 queue:Reg/w0:  redacted:20514/RELP
4298.010296961:action 0 queue:Reg/w0: relpSessWaitState returns 0
4298.010304564:action 0 queue:Reg/w0: frame to send: '4 syslog 1531 redacted {"file":"/logs/apps/app/app.log","msg":"INFO class:000 - User - Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum. Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum. Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum."}
'
4298.010349765:action 0 queue:Reg/w0: relpTcpSend: sock 8, lenbuf 1546, send returned 1546 [errno 11]
4298.010356591:action 0 queue:Reg/w0: ADD sess 0x7f6890003e90 unacked 3, sessState 4
4298.010360865:action 0 queue:Reg/w0: sendbuf added to unacked list
4298.010366716:action 0 queue:Reg/w0: omrelp: shutdownImmediate ptr now is 0x1a81bd0
4298.010401185:action 0 queue:Reg/w0: entering actionCalldoAction(), state: itx, actionNbr 0
4298.010406347:action 0 queue:Reg/w0:  redacted:20514/RELP
4298.010411140:action 0 queue:Reg/w0: relpSessWaitState returns 0
4298.010424968:action 0 queue:Reg/w0: frame to send: '5 syslog 1978 redacted {"file":"/logs/apps/app/app.log","msg":"INFO class:000 - User - Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum. Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum. Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum. Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum."}
'
4298.010466728:action 0 queue:Reg/w0: relpTcpSend: sock 8, lenbuf 1993, send returned 1993 [errno 11]
4298.010472974:action 0 queue:Reg/w0: ADD sess 0x7f6890003e90 unacked 4, sessState 4
4298.010477234:action 0 queue:Reg/w0: sendbuf added to unacked list
4298.010483807:action 0 queue:Reg/w0: omrelp: endTransaction
4298.010489663:action 0 queue:Reg/w0: Action 0 transitioned to state: rdy
4298.010494491:action 0 queue:Reg/w0: actionCommit, action 0, in retry loop, iRet 0
4298.010499761:action 0 queue:Reg/w0: regular consumer finished, iret=0, szlog 0 sz phys 3
4298.010507657:action 0 queue:Reg/w0: queue.c:1947:ConsumerReg:invocation 4: WARNING: mutex still owned by us as we exit function, mutex:  0x1a87fa0[0/action 0 queue:Reg/w0 (7f689ac8e700)]
4298.010521776:action 0 queue:Reg/w0: DeleteProcessedBatch: we deleted 3 objects and enqueued 0 objects
4298.010527826:action 0 queue:Reg/w0: doDeleteBatch: delete batch from store, new sizes: log 0, phys 0
4298.010533847:action 0 queue:Reg/w0: regular consumer finished, iret=4, szlog 0 sz phys 0
4298.010545310:action 0 queue:Reg/w0: queue.c:1947:ConsumerReg:invocation 4: WARNING: mutex still owned by us as we exit function, mutex:  0x1a87fa0[0/action 0 queue:Reg/w0 (7f689ac8e700)]
4298.010551016:action 0 queue:Reg/w0: action 0 queue:Reg/w0: worker IDLE, waiting for work.
...
```

Probably related to: https://github.com/rsyslog/rsyslog/issues/523

**Comment by davidelang** [comment]
in your config is readmode set to the default, and does the last line of the file have a newline?

**Comment by mostolog** [comment]
Log file has a newline at the end.

> input(type="imfile" file="/logs/apps/app/app.log" tag="mytag" addMetadata="on" ruleset="json" startmsg.regex="^[[:digit:]]{2} [[:alpha:]]{3} [[:digit:]]{4}" PersistStateInterval="1")

As I'm on 8.22, still not using readTimeout (>= 8.23). There's a system timeout I should wait for?

**Comment by davidelang** [comment]
imfile doesn't know that it's at the end of a line until it detects the beginning of the next line.

with readmode = 0 and no startmsg.regex it can do this when it sees the newline. In all other cases it needs to see the beginning of the next line to know that the prior line has finished.

There has been some discussion of adding a timeout to say that if nothing new in the file is detected for some time, consider it the end of a message and send it. But this has not been implemented yet (and if it was, it would result in some 'odd' cases where what gets send doesn't look like the file if writes don't properly terminate a message)

**Comment by mostolog** [comment]
IIUC, you mean that, when using _startmsg.regex_, that line won't be flushed until:
 - file is deleted
 - file is renamed
 - if new (**starting with regex**) line is readed, old line will be flushed, and current will be buffered
 - if new (**NOT starting with regex**) line is readed, this will be appended to the current buffer

Will 8.23 readTimeout change this behavior (and lines will be flushed after timeout)?

**Comment by davidelang** [comment]
imfile reads data in a character at a time, when it detects the end of a message, it sends it.

for the traditional stuff, the end of a line indicated the end of a message and it would send it (and if the file is deleted/renamed, it continues to read to the end of the file, and then sends any fragment that it has)

with a regex or readmode != 0 it doesn't know that a message has ended until it can read enough of the next message to detect that it is a new message

I believe that the 8.23 readTimeout is intended to address this issue by sending what's been buffered if there is no new data for long enough. (since I've been catching up and not following everything, I'm only p

**Comment by mostolog** [comment]
I'll test with 8.23 and let you know. By now, this issue can be tagged as "need more info".

Anyway, IMHO adding a default 30/60s timeout to flush events using a configuration setting will be better than current behavior.

**Comment by rgerhards** [comment]
Adding the timeout has performance implications, so I don't think this
would be good as default. The new feature is already documented.

Rainer

Sent from phone, thus brief.

Am 07.11.2016 13:40 schrieb "mostolog" <notifications@github.com>:

**Comment by mostolog** [comment]
This seems to be properly solved with 8.23 and readTimeout
